src/core/sock/sock-redirect.cpp

```diff
@@ -3171,14 +3243,110 @@ extern "C" EXPORT_SYMBOL pid_t waitpid(pid_t pid, int *wstatus, int options)
      *     * NGINX at some future point forks a new worker process(es) to replenish the worker
      * process tally.
      */
-    if (safe_mce_sys().actual_nginx_workers_num > 0 && child_pid > 0 && !WIFCONTINUED(*wstatus)) {
-        auto worker_pid =
-            std::find(g_p_nginx_worker_pids->begin(), g_p_nginx_worker_pids->end(), child_pid);
-        if (worker_pid != g_p_nginx_worker_pids->end()) {
-            *worker_pid = -1;
-        }
+    if (g_p_app && g_p_app->type == APP_NGINX && child_pid > 0 && !WIFCONTINUED(*wstatus)) {
+        std::lock_guard<decltype(g_p_app->m_lock)> lock(g_p_app->m_lock);
+        g_p_app->unused_worker_id.insert(g_p_app->get_worker_id());
+        g_p_app->map_thread_id.erase(getpid());
     }
     return child_pid;
 }
 
+static int proc_nginx(void)
+{
+    int rc = 0;
+
+    DO_GLOBAL_CTORS();
+    std::lock_guard<decltype(g_p_app->m_lock)> lock(g_p_app->m_lock);
+
+    auto itr = g_p_app->map_listen_fd.begin();
+    while (itr != g_p_app->map_listen_fd.end()) {
+        int listen_fd = itr->first;
+        itr++;
+        rc = init_worker(g_p_app->get_worker_id(), listen_fd);
+        if (rc != 0) {
+            srdr_logerr("Failed to initialize worker %d, (errno=%d %m)", g_p_app->get_worker_id(),
+                        errno);
+            break;
+        }
+    }
+
+    return rc;
+}
+
+#if defined(DEFINED_ENVOY)
+static int proc_envoy(int __op, int __fd)
```

Comment:
Please move nginx/envoy specific logic to a separate file: nginx_envoy.cpp or something like that